### ISSUE TITLE
videoio: enum CAP_MODE_* should be removed

### ISSUE THREAD

**Issue by paroj** [opencv/opencv#5485]
`CAP_MODE_*` enum, is supposed to be used as `set(CAP_PROP_MODE, CAP_MODE_YUYV)`.

However this makes `CAP_PROP_MODE` ambiguous:
- the `cap_libv4l` implementation interprets it as "output mode" and  returns YUYV images.
- the `cap_*1394*` implementations interpret it as "transmission mode" and always return BGR images. (even converting gray to BGR)
- the `cap_v4l` implementation also interprets it as "transmission mode", but treats it read only
- other `cap_*` implementations do not support `CAP_PROP_MODE`

looking at the documentation `CAP_PROP_MODE` we find:

> Backend-specific value indicating the current capture mode.

therefore I think the `cap_libv4l` behavior wrong:
1. the call should result in YUYV transmission frames, but still return BGR
2. the currently available API for selecting output format is `CAP_PROP_CONVERT_RGB`, which allows to disable conversion and get the transmission format
3. the user actually should not be able to specify the output format at all - even though it may save a memcpy
   1. this would mean we have to move half of the `cvtColor` API to VideoCapture. `CAP_MODE_YUYV` implies `COLOR_XXXX2YUYV` - so `CAP_MODE_Lab` would make sense as well
   2. lets consider the transmission format is Bayer and the output is RGB. Which demosaicing algorithm should be used? Edge Aware, VNG, the fastest? 
   3. Lets consider the user wants GRAY images and the possible transmission formats are MJPEG and YUYV. Which one would be more efficient for VideoCapture to select?

allowing to specify the output format requires a large amount of logic inside VideoCapture which does not belong there.

However there are advantages in adding a API similar to `cv::IMREAD_GRAYSCALE` to VideoCapture - yet the `CAP_MODE_*` values are the wrong way to do so.

I suggest removing or deprecating the `CAP_MODE_*` enum as long as `cap_libv4l` is still the only implementation understanding it.

**Comment by JoeHowse** [comment]
The point of `CAP_PROP_MODE` is to provide optimized capture of frames in a *known, specified format*.  Importantly, when using this API, the caller does not need to know the camera's capture format, and `set()` may reconfigure the capture format to a better optimized format.  Depending on the camera, the format conversion may be performed in the camera's firmware, which will be a better optimized approach than doing a conversation later in software.

Note that `CAP_PROP_CONVERT_RGB` is unsupported in most implementations.  I think the DC1934 and unicap implementations may support it, but nothing else does.

The videoio module contains many capture properties that are supported by some implementations but not others.  I would like to see `CAP_PROP_MODE` supported in more implementations, not removed.  At present, the libv4l support is especially valuable on Linux systems with constrained resources, such as single-board computers, where you would want to minimize the size of input. For example, `CAP_MODE_GRAY` may configure the camera in such a way that input size is cut in half (12 bpp YUV instead of 24 bpp RGB).

**Comment by paroj** [comment]
you actually have a point when your camera supports several pixelformats in hardware. Still I am yet to see a camera where you can choose whether you want RGB, BGR or YUV. Typically you have one *native* format and one or more compressed ones. The conversion to BGR is then made in software (inside libv4l or opencv).

My point is that you can achieve your goal of a smaller input size without a new API by telling the driver to give you the raw input (`CAP_PROP_CONVERT_RGB=0`) and then doing the color conversion manually.

If the native format is YUV you can call `cvtColor(CV_YUV2GRAY)` to extract the Y channel (like what you did inside libv4l).
If the format is MJPEG you can do the same during decoding, but you get even smaller bpp.
If the format is Bayer, you need to figure out what is appropriate..

While you surely could hide all the complexity behind `CAP_MODE_GRAY`, I think it does not belong inside the VideoCapture.

**Comment by JoeHowse** [comment]
Many cameras do support multiple pixel formats in firmware, so the API's advantages will generalize well to future implementations.  For example, Point Grey Research cameras typically support Bayer, RGB8, Mono8, Mono16, and other formats in firmware.  Some webcams, such as PS3 Eye, support both Bayer and YUV formats in firmware (though the user's system may or may not have drivers that can take advantage of the firmware's support for both formats).

With libv4l, format conversions may run in the camera's firmware, or may run in the host's kernel-space software, which is arguably still better than running in user-space software.

videoio's default behavior is to hide complexities of the camera's native format and just convert to BGR.  CAP_PROP_MODE follows the same philosophy but allows the caller to specify the destination format (BGR or another).

Your proposed alternative has these limitations:
* The videoio implementation must know which of the camera's supported formats is "native". This is not necessarily the same as the camera's default or current format.
* The caller must know the camera's native format.
* Conversions are always performed in user-space software.
* cvtColor must support the appropriate conversion.

**Comment by paroj** [comment]
Note that I am not arguing against `CAP_PROP_MODE` - you still could set it do a backend specific value to configure the pixelformat. (I did not think about it in my last reply)

What I am arguing against are the `CAP_MODE_*` values like `CAP_MODE_GRAY` because it requires the videoio implementation to know the best "native" format to match this request. This is actually your first point against my approach - yet it is exactly what it tries to solve.

What my approach actually requires is
> The caller must know the camera's native format.

because this is ambiguous - think of a case where both `YUV444` and `YUV411` are supported. Note that finding out the *currently used* format is as easy as `get(CAP_PROP_MODE)`.

>  Conversions are always performed in user-space software.

when the camera firmware does not support the desired pixelformat, they have to happen *somewhere*. In case of *libv4l* it [also happens in userspace](https://github.com/philips/libv4l/blob/master/libv4lconvert/rgbyuv.c#L93) - it is abstracted away from us. Using `cvtColor` has the advantage of a more tested and more optimized implementation (openCL, SIMD, threaded). BTW doing stuff in kernel-space does not make it magically faster.

> cvtColor must support the appropriate conversion.

if you could get a RGB image before, there must be some code that did the conversion - be it `cvtColor` or something else. If you explicitly say `CAP_PROP_CONVERT_RGB=0` it probably means that you now want to do a different conversion (manually).

**Comment by JoeHowse** [comment]
If `cvtColor` is the better optimized conversion in some cases, then the implementation of videoio can use `cvtColor` in those cases. In `cap_libv4l.cpp`'s current support for `CAP_MODE_GRAY`, conversion from planar YUV to Y8 is achieved by just `memcpy`ing the Y plane, and I think `cvtColor` would *not* be better optimized in such cases. But these are implementation details.

More importantly, note that typical implementations of videoio's capture functionality will return a *copied* buffer rather than the original input buffer. If the caller just wants the Y data, it would be suboptimal to return a copy of all the YUV data. We would be copying at least 1.5x the required data in such cases.

To support reasonable optimizations in its input handling, videoio needs an API that allows the caller to specify a *destination* format (such as Y8 gray). If `CAP_PROP_MODE` is not the right API for this, we need some other API for it.

**Comment by paroj** [comment]
following your feedback, I went ahead and implemented `CAP_PROP_CONVERT_RGB` for `cap_v4l.cpp` https://github.com/paroj/opencv/commit/d69e48ad4e50f65148c36d9f7dd5431cbe7f0ba9

giving you the raw v4l buffer when disabled. Do you think this will work for you?

**Comment by JoeHowse** [comment]
I understand the motivation, but I have concerns about the safety of returning a `cv::Mat` that wraps a raw capture buffer.  Similar behavior was present in OpenCV 1.x (using an `IplImage *`) with the following warning in the [documentation](http://docs.opencv.org/modules/highgui/doc/reading_and_writing_images_and_video.html#videocapture-retrieve):

> OpenCV 1.x functions cvRetrieveFrame and cv.RetrieveFrame return image stored inside the video capturing structure. It is not allowed to modify or release the image! You can copy the frame using cvCloneImage() and then do whatever you want with the copy.

By returning a `Mat` that wraps a raw capture buffer, we create the following side-effects, I believe:

* The `Mat`'s data will be modified in subsequent calls to `VideoCapture::grab` or `VideoCapture::read`.
* The `Mat`'s data will be freed in `VideoCapture::release` or `~VideoCapture`.
* The `Mat` may free its data before the VideoCapture is `release`d or deleted, and in this case the capture gets into a bad state.

At minimum, these issues would need to be documented. However, I am not convinced that `VideoCapture::grab` or `VideoCapture::read` should *ever* return a `Mat` with unsafe data. Maybe others can weigh in with opinions on this point.

**Comment by paroj** [comment]
it actually returns a `IplImage*` and not a `cv::Mat`. It is copied to a Mat in `VideoCapture::read`. 
So actually it just reduced the transferred data from 4x to 2x (where 1x is just copying the Y channel).
To get the raw pointer (1x copy/ no copy) one would still need to use `cvRetrieveFrame`.

Still I think we are now at a reasonable level of optimization. `CAP_MODE_GRAY` is probably the only case where we could get better using the`CAP_MODE_*` API.
For BGR, RGB, and YUYV there are no advantages of specifying a destination format.

**Comment by JoeHowse** [comment]
> it actually returns a IplImage* and not a cv::Mat. It is copied to a Mat in VideoCapture::read.

Thanks for pointing this out. This may allow for opportunities to remove other copying from the lower-level parts of the implementations. I will check it out.

> CAP_MODE_GRAY is probably the only case where we could get better using theCAP_MODE_* API.
For BGR, RGB, and YUYV there are no advantages of specifying a destination format.

Y8 grayscale is perhaps the most commonly used format in computer vision, so it is an important case. Moreover, I believe that the advantages of specifying a destination format are also relevant to other formats (RGB, etc.). Consider the following example.

A user wants RGB data (not BGR). The camera's native format is some Bayer format. In Strategy A, which I believe is what you propose:

1. The videoio module captures Bayer data to a buffer.
2. The videoio module copies Bayer data to a Mat, which is returned to the user.
3. The user queries videoio to find out what format was received.
4. The user converts the Bayer Mat to a destination RGB Mat (probably in a switch statement, covering all the possible formats that might be received).

Why did we copy Bayer data in step 2? This is extraneous copying because, after all, the user did not want the Bayer data, and just has to convert it in a later step. And in step 4, does the user even know how to convert this particular Bayer format to RGB efficiently and correctly? If not, will OpenCV provide optimized and tested cvtColor implementations for every possible Bayer format? Normally, this would be the responsible of the camera firmware developer or driver developer.

In Strategy B, which I propose:

1. Preferably, Bayer-to-RGB conversion occurs in some way that was optimized and tested by the camera firmware developer or the driver developer.
2. The videioio module captures RGB data to a buffer.
3. The videoio module copies RGB data to a Mat, which is returned to the user. No further conversion is necessary.

Strategy B is certainly more user-friendly for a user who just wants an RGB frame (or Y8 gray frame, etc.), it doesn't explicitly copy unwanted data, and it potentially takes advantage of better-optimized and better-tested conversion code written by the camera's firmware developer or driver developer.

We seem to be somewhat at cross-purposes in terms of our criteria for a capture strategy, so maybe others can weigh in with their opinions.

**Comment by StevenPuttemans** [comment]
At the request of @JoeHowse I took the liberty of reading all this. Not that everything makes sense to me, but I am afraid that this will be a discussion of like and dislike in flavors on how to set camera interfaces.

It is well known that the OpenCV camera capture interface is far from straightforward. A given constant is only implemented for specific interfaces, and not supported for the majority of things.

Personally I think the only way to even start discussing individual parameter changes that were introduced to enable some functionality, like in this case, would be to completely clean up the interface, make every cap interface aware of all properties and implement relevant handlers for them in the designated capture interfaces. However this is a huge work ...

As some sidenotes
- I agree with @JoeHowse that Y8 grayscale is perhaps the most commonly used format in computer vision and that there are many camera's out there supporting the output of different pixel configurations straight from the camera itself. That will always be more effective than pulling raw data and processing it on your machine, especially when you are working with embedded devices with limited resources.
- At the point of *A user wants RGB data (not BGR)* I do not agree. A user using OpenCV should be aware from the start that the consensus in OpenCV is to use BGR data. That is one of the main reasons why I literally hate all the conversions to RGB, because it screws up the perceptions of newbies, leading to even more troubles than to just accept that BGR is the main OpenCV format. This is the same for many other computer vision libraries out there.
- *Strategy B is certainly more user-friendly for a user who just wants an RGB frame* --> I agree here, again keeping the embedded field in mind.

We cannot forget that one of the main focusses of OpenCV is to provided embedded computer vision solutions at decent speed and optimizations, thus the more processing that is being performed on the dedicated memory of the camera itself, the better in my opinion.

**Comment by paroj** [comment]
I kind of agree that VideoCapture should additionally offer efficient grayscale image retrieval - maybe similar to `cv::IMREAD_GRAYSCALE` - but that should be tracked in a different issue.

This one is about the `CAP_MODE_*` values. Note that this is neither about removing `CAP_PROP_MODE`.

The user still can do `set(CAP_PROP_MODE, V4L2_PIX_FMT_H264)` to specify a backend specific transmission format.

What the user should not be able to do is `set(CAP_PROP_MODE, CAP_MODE_YUYV)` and expect a YUYV image to pop out. Why?

1. capture mode and output format are two different things
  1. the line above means: transmit frames using YUYV and give me BGR (as opposed to MJPEG transmission)
  2. there is currently not really an API for selecting output format besides `CAP_PROP_CONVERT_RGB`. (which just allows saying: give me images in the transmission format)
2. the user should not be able to specify the output format at all:
  1. this would mean we have to move half of the `cvtColor` API to VideoConvert. `CAP_MODE_YUYV` implies `COLOR_XXXX2YUYV` - why not `CAP_MODE_Lab`? After all I might want an Lab image and this would save one copy there as well.
  2. lets consider the transmission format is Bayer and the output is RGB. Which demosaicing algorithm should be used? Edge Aware, VNG, the fastest? 
  3. Lets consider the user wants GRAY images and the possible transmission formats are MJPEG and YUYV. Which one would be more efficient for VideoCapture to select?

basically allowing to specify the output format requires a large amount of logic inside VideoCapture which does not belong there.

some other notes:
> Why did we copy Bayer data in step 2?

I would call this an API limitation preventing you to shoot yourself in the foot - you could use the C API to avoid it after all..

> Preferably, Bayer-to-RGB conversion occurs in some way that was optimized and tested by the camera firmware developer or the driver developer.

I think this is a poor assumption. If you request a format not supported by the firmware, the driver will have to do it in software. And I really doubt that this "fallback" implementation will be better than cvtColor using IPP or tegra optimizations.

**Comment by mself** [comment]
FYI, there was some further discussion on this topic in #6678.

**Comment by paroj** [comment]
as discussed in #6678, the only sensible option for `CAP_PROP_MODE` is `CAP_MODE_GRAY`.
To disambiguate `CAP_PROP_MODE` I would therefore suggest to introduce a boolean `CAP_PROP_MODE_GRAYSCALE` property instead.

**Comment by mself** [comment]
@JoeHowse, @StevenPuttemans, this seems reasonable to me.  Do we have consensus on this?  If so, I can create the `CAP_PROP_MODE_GRAYSCALE` property in the AVF capture class.

Regarding the name, should it be `CAP_PROP_CONVERT_GRAYSCALE` instead (to match `CAP_PROP_CONVERT_RGB`)?  I'm not actually clear what `CAP_PROP_CONVERT_RGB` does -- what format will you get if you set it to false?  Is it just device/file specific?  The docs don't say.

or maybe just `CAP_PROP_GRAYSCALE`?

**Comment by JoeHowse** [comment]
What would it mean if both CAP_PROP_CONVERT_GRAYSCALE and CAP_PROP_CONVERT_RGB are true?  This API allows the user to specify meaningless combinations, and it becomes increasingly awkward if more conversion modes are added.

Maybe something like CAP_PROP_CONVERSION would make more sense, with possible values defined by an enum such as CAP_CONVERSION { BGR, NONE, GRAYSCALE } or CAP_CONVERSION { BGR, NONE, GRAYSCALE, RGB, YUVY }.  I am still in favor of supporting RGB and YUVY, as many cameras support these conversions efficiently in firmware and many users may want them for interoperability with other graphics or video frameworks.  Like other features of the VideoCapture API, they do not have to be supported by all back-ends.

NONE would be no conversion, i.e. whatever format the camera provides in its current capture mode.

As a side-effect, the CAP_PROP_CONVERSION setter should be permitted to change CAP_PROP_MODE for efficiency.

**Comment by paroj** [comment]
so how about:
- `CAP_PROP_CONVERT` with values
  - `CAP_CONVERT_BGR` (default)
  - `CAP_CONVERT_GRAY`
  - `CAP_CONVERT_OFF`

add deprecation comment to
- `CAP_PROP_CONVERT_RGB`
- `CAP_MODE_*` enum

to get YUV or RGB images the user would set `CAP_CONVERT_DISABLE` and set `CAP_PROP_MODE` appropriately.
In contrast to grayscale where you only have to copy 1/2 or 1/3 of the data you gain nothing by special handling inside VideoCapture for those.

**Comment by JoeHowse** [comment]
> -  CAP_PROP_CONVERT with values
>  -  CAP_CONVERT_BGR (default)
>  - CAP_CONVERT_GRAY
>  - CAP_CONVERT_OFF

Okay, let's proceed based on this proposal.  If the community needs other conversions in the future, we can add them to the `CAP_CONVERT` enum.

For `CAP_PROP_MODE`, it would be nice if OpenCV provided some way to query the supported modes so that a user could try to find a YUV mode, an RGB mode, or whatever in a manner that is somewhat independent of the backend.  However, I guess this is a separate issue.
